Improve tutorial notebooks

OriolAbril opened the pull request:
Pages to review:
* Build custom plot with PlotCollection (new): https://arviz-plots--261.org.readthedocs.build/en/261/tutorials/compose_own_plot.html
* Intro to batteries-included plots (updated): https://arviz-plots--261.org.readthedocs.build/en/261/tutorials/plots_intro.html
* Using __existing__ PlotCollection objects (updated): https://arviz-plots--261.org.readthedocs.build/en/261/tutorials/intro_to_plotcollection.html


<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--261.org.readthedocs.build/en/261/

<!-- readthedocs-preview arviz-plots end -->

OriolAbril commented:
I plan to add an [explanation](https://diataxis.fr/explanation/) page about handling of aesthetics but I think the 3 notebooks already present are in good shape for review. We could also merge this and make another PR

fonnesbeck commented:
No ReviewNB? I will just quote text then.

fonnesbeck commented:
> plotting functions that generate the minimal combination of artists

Will most users know what "artists" are?

aloctavodia commented:
I just activated ReviewNB on this repo

tomicapretto commented:
@OriolAbril, I added myself as a reviewer because I want to have a look at these new things :D

tomicapretto commented on docs/source/tutorials/plots_intro.ipynb:
```diff
@@ -530,6 +530,7 @@
    "id": "5211f6de-d447-46a8-8652-3c9fd77d7915",
```
<p>, <em>and</em> ,&nbsp;<code>ci_kind</code>&nbsp;and&nbsp;<code>ci_prob</code>&nbsp;to control the credible interval. </p>

tomicapretto commented on docs/source/tutorials/plots_intro.ipynb:
```diff
@@ -530,6 +530,7 @@
    "id": "5211f6de-d447-46a8-8652-3c9fd77d7915",
```
<p>I think it will help to explain with an example what it means to "reduce the sampled dimensions" and/or what arviz understands for "sampled dimensions"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

tomicapretto replied:
<p>I think it would help to display <code>schools</code> before passing it to the <code>plot_dist</code> function  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

tomicapretto replied:
<p>This is stronger than me, I correct this to my students all the time, so I also have to do it here :)</p><p> </p><p>"A probability density, here represented with a Kernel Density Estimate (KDE)" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

tomicapretto replied:
<blockquote>&nbsp;to choose how to represent the probability density</blockquote><p>Is it always a probability density? Could we use "the probability distribution"? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

tomicapretto replied:
<p>Make it more explicit that you refer to keys in the dictionary you pass to <code>plot_kwargs</code>  . When I looked at the docs I was looking for arguments in the function itself.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

tomicapretto replied:
<p>KDE in uppercase</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

tomicapretto replied:
<p>Isn't it the first element with the empty symbol?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

tomicapretto replied:
<p>What do you mean with "we have called properties in which we encode aesthetics"? I don't understand that part.</p><p> </p><p>Also, not sure if some separation in the form of a comma or period is needed in the last sentence (function otherwise)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

tomicapretto replied:
<blockquote>Combining aesthetics applied to variables and to dimensions we can generate a plot to compare the global parameters to the group parameters when we have a hierarchical model as in general the global parameter will be a scalar and the group ones a multidimensional one</blockquote><p>Looks like a very long sentence :D</p><p> </p><p>Do you want to have a comma before the "How close"...?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

tomicapretto replied:
<pre class="ql-syntax" spellcheck="false">Line #16.        aes_map={"point_estimate": ["color", "overlay", "alpha"]},
</pre><p>I'm struggling to understand this. But I confess I have not read this very thoroughly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

jessegrabowski replied:
<p>Piling on to commentary about <code>plot_kwargs</code> -- I would not expect the keys here to be graphical elements. Typically, a function's <code>_kwargs</code>are forwarded as-is somewhere, but this is giving finer control than that. I think it requires some further clarification on this point. Because it's breaking a typical pattern, I would even consider copying the docstring into this notebook cell, rather than just linking to it, so you can really break down what is going on here.</p><p> </p><p>Also you need an oxford comma here too : )</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

jessegrabowski replied:
<p>Nit: Start the sentence with "for example" -&gt; "For example, we can use it in <code>plot_dist</code> to control..." </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

jessegrabowski replied:
<blockquote>Some other graphical elements for <code>plot_dist</code> are <code>point_estimate</code>, <code>point_estimate_text</code>, <code>title</code>, etc</blockquote><p>Isn't that the exhaustive list (along with <code>kde</code>, which was already mentioned) ? "etc" doesn't make sense here if so, and the sentence should read:</p><p> </p><p>"The other graphical elements for <code>plot_dist</code>are ...</p><p> </p><p>Comment 2:</p><p> </p><p>NIT: Comma after for instance: "For instance, to map the dataset..."</p><p> </p><p>Comment 3:</p><p> </p><p>NIT: Don't start sentences with contractions: "<s>And</s> To manually set the figure size..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

jessegrabowski replied:
<p>Maybe clearer:</p><p> </p><p>So far, we have looked at examples where we adjusted the color and linestyle properties of the plotting backend using<code>pc_kwags</code> and <code>aes_map</code>.  Next, we will see how to use these arguments to do much more complex adjustments. </p><p> </p><p>When adjusting the aesthetics of a graphical element, the only limitation is that the keys and values are valid passed to a graphical property are valid for the specific backend function being used. As long as this is satisfied, anything goes!</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

jessegrabowski replied:
<p>Suggestion:</p><p> </p><p>When working with hierarchical models, it is common to want to compare how the distribution of each unit (each school, in the example we've been working with) compares to the population distribution implied by their shared hyperparameters. In general, these hyperparameters will be scalars, while the unit effects will have a dimension (e.g. school). </p><p> </p><p>In the next plot, we use <code>pc_kwargs</code> and <code>aes_map</code>  to overlay the population mean <code>mu</code> on the each school's unique effect <code>theta</code> . </p><p> </p><p> </p><p>&lt;/End suggestion, begin comment&gt;</p><p>I'm pretty sure these don't actually compare the same thing though. Distribution over the population of effects is <code>N(mu, tau)</code>. So just plotting theta against mu discards the estimated variance of the population of school effects?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

jessegrabowski replied:
<p>This is a good point to make, but we could also pass <code>plot_kwargs = {'kde': {'zorder': 100}}</code> and somehow only apply that to <code>mu</code> via <code>pc_kwargs</code> I guess ? Might be a useful exercise to show, since after having read this, it's not immediately obvious to me how to do it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

jessegrabowski replied:
<blockquote>The functions used by <code>plot_dist</code> (and by other <code>plot_...</code> functions) aim to be somewhat general, so for example, <code>y</code> is a valid key for encoding information</blockquote><p>Run-on sentence. Split as: ...aim to be somewhat general. For example, <code>y</code> .... </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

jessegrabowski replied:
<p>Run-on sentence. Suggestion:</p><p> </p><p>We can have a range of variables with different shapes and dimensions,. As long as we are careful that the facetting and mapping arguments do not conflict, the underlying <code>PlotCollection</code> used by <code>plot_dict</code> can combine all the different variables into a single plot. The provided aesthetics will be mapped only when relevant.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

jessegrabowski replied:
<p>Add an explanation between the <code>Dataset</code> cell and the plotting code explaining all the choices that are being made. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

tomicapretto commented on docs/source/tutorials/plots_intro.ipynb:
```diff
@@ -530,6 +530,7 @@
    "id": "5211f6de-d447-46a8-8652-3c9fd77d7915",
```
<pre class="ql-syntax" spellcheck="false">Line #11.            "aes": {"color": ["__variable__"], "overlay": ["school"], "alpha": ["__variable__"]},
</pre><p>Would be good to have a comment explaining why we use <code>__variable__</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

jessegrabowski replied:
<p>+1</p>

aloctavodia commented on docs/source/tutorials/plots_intro.ipynb:
```diff
@@ -530,6 +530,7 @@
    "id": "5211f6de-d447-46a8-8652-3c9fd77d7915",
```
<p>I think we are only mentioning the method associated to style, but not the style themselves<a href="https://arviz-plots.readthedocs.io/en/latest/api/helpers.html#style" target="_blank">https://arviz-plots.readthedocs.io/en/latest/api/helpers.html#style</a> styles are listed here <a href="https://arviz-devs.github.io/EABM/Chapters/Elements_of_visualization.html#style-sheets" target="_blank">https://arviz-devs.github.io/EABM/Chapters/Elements_of_visualization.html#style-sheets</a>  (but this is outside the docs)</p>

OriolAbril commented:
<p>This is one of the multiple reasons I don't like reviewnb. This is hidden from the rendered version. I took the example model+data from Ben's post <a href="https://drbenvincent.github.io/posts/mcmc_grammar_of_graphics.html" target="_blank">https://drbenvincent.github.io/posts/mcmc_grammar_of_graphics.html</a> as it seemed like a simple enough case to focus on PlotCollection use while also complex enough to allow multiple coherent and common visualization elements as well as faceting and aesthetic mapping combinations. At some point it will ideally be <code>load_arviz_data</code> with this or a similar dataset though. </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/#comment-2912059135'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

OriolAbril commented:
<p>I thought about this initially but then thought showing empty plots might be a bit sad, I'll try it and see how it goes. Then instead of the full examples being here they can be in each of the option1/2 sections.</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/#comment-2912059299'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

OriolAbril commented:
Thanks for all the feedback! I have started addressing the comments on `compose_own_plot`, I need to reorganize the plotcollection initialization and function definition, but I think I addressed everything else. I have also opened a couple issues of improvements after reading the comments and @aloctavodia opened https://github.com/arviz-devs/arviz-plots/pull/269.

We currently have a `seealso` we generally have at the bottom with other pages they might want to check out next. I was thinking we could add an "info" or "audience" box at the top where we explain what is the audience we had in mind when writing that page. My idea so far:

* `compose_own_plot` -> ideally not many assumptions about familiarity with arviz-base beforehand, maybe having gone over the overview page only? but they have a clear understanding of bayesian modeling and want/need something out of the batteries included plots.
* `intro_plotcollection` -> Intermediate users of arviz-plots who are already comfortable using batteries included plots and who want to make some extra tweaks afterwards
* Intro to batteries included plots -> has gone over the overview and is already familiar with batteries included plots but wants to make the most out of them. Now that we have the overview I will make it a bit more advanced. It is currently a bit over the place and tries to cover too much

Do you think seeing this (with better wording) at the top of the pages would help?

jessegrabowski commented:
<p>Yes it was a comment <code>obs_idx</code> , needing to put a dummy dimension on the data index, when in reality this should have a pandas multiindex of <code>(group, observation)</code> . </p><p> </p><p>It was a comment unrelated to the content of these nbs</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/#comment-2912059166'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

OriolAbril commented:
<p>We use plot or target in general for this, <code>ax</code>/axes is the matplotlib equivalent. Naming is a bit of a nightmare when taking into account the different backend existing conventions: <a href="https://arviz-plots.readthedocs.io/en/latest/glossary.html#equivalences-with-library-specific-objects" target="_blank">https://arviz-plots.readthedocs.io/en/latest/glossary.html#equivalences-with-library-specific-objects</a>.</p><p> </p><p>I wanted to keep this example as simple as possible and refer them to the other notebook, but maybe this plotting operation is too dumb. In general we can't automatically wrap the incoming functions, they will have more complicated signatures, and might even require some processing of the input arguments before calling the plotting backend functions.</p><p> </p><p>I am not even sure how we'd do it here. <code>PlotCollection</code> calls functions of the style <code>fn(da, target, **kwargs)</code> . Say we allow a string as first positional argument too. We could automatically wrap it to do <code>getattr(target, fn_str)(da, **kwargs)</code> but that isn't what we want here, we ignore the <code>da</code> argument and hardcode the line at 0    </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/arviz-plots/pull/261/discussion/#comment-2912059866'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

OriolAbril commented:
I think I have addressed all the comments. Thanks so much for the in depth reviews, the notebooks look much better now!
